File: alpha/experiments/item-storage-protocol/src/receipts.ts
// Transfer receipts: postage by weight, not by trips, signed on each end.
//
// Bytes move in fixed-size increments. After each increment the receiver signs
// an acknowledgment (direction, item, byte range, running total, timestamp) and
// the sender countersigns. Both parties append the same co-signed receipt to
// their own ledgers, so the two ledgers reconcile to identical totals and
// neither side can quietly restate the weight moved.
//
// "Meter the boundary, not the machine": billing is the sum of signed acks, not
// a claim about internal work.

import type { Json } from "./canonical.ts";
import type { Actor } from "./actor.ts";
import { verifyJson } from "./crypto.ts";

export type Direction = "upload" | "download";

export interface Ack {
  direction: Direction;
  cid: string;
  rangeStart: number; // inclusive
  rangeEnd: number; // exclusive
  incBytes: number;
  runningTotal: number; // cumulative bytes for THIS transfer
  ts: number;
}

export interface Receipt {
  ack: Ack;
  sender: string;
  receiver: string;
  senderSig: string;
  /** null models a walkaway: the receiver took the increment but never signed. */
  receiverSig: string | null;
}

function ackCore(a: Ack): Json {
  return {
    direction: a.direction,
    cid: a.cid,
    rangeStart: a.rangeStart,
    rangeEnd: a.rangeEnd,
    incBytes: a.incBytes,
    runningTotal: a.runningTotal,
    ts: a.ts,
  };
}

export interface TransferResult {
  receipts: Receipt[];
  total: number;
  /** Set when the receiver walked away on the final increment. */
  unsignedExposureBytes: number;
}

export interface TransferOptions {
  /** If true, the receiver takes the final increment without signing it. */
  walkawayOnLast?: boolean;
}

/**
 * Run one metered transfer. The receiver signs each increment; the sender
 * countersigns; both append the co-signed receipt to their ledgers. Returns the
 * receipts and the reconciled total.
 */
export function runTransfer(
  direction: Direction,
  cid: string,
  totalBytes: number,
  incrementSize: number,
  sender: Actor,
  receiver: Actor,
  ts: number,
  opts: TransferOptions = {},
): TransferResult {
  const receipts: Receipt[] = [];
  let moved = 0;
  let unsignedExposureBytes = 0;

  const nIncrements = Math.ceil(totalBytes / incrementSize);
  for (let i = 0; i < nIncrements; i++) {
    const start = i * incrementSize;
    const end = Math.min(start + incrementSize, totalBytes);
    const incBytes = end - start;
    moved += incBytes;
    const ack: Ack = {
      direction,
      cid,
      rangeStart: start,
      rangeEnd: end,
      incBytes,
      runningTotal: moved,
      ts,
    };

    const isLast = i === nIncrements - 1;
    const walkaway = Boolean(opts.walkawayOnLast) && isLast;

    // The receiver signs first (they acknowledge receipt); the sender
    // countersigns. On a walkaway the receiver never signs.
    const receiverSig = walkaway ? null : receiver.sign(ackCore(ack));
    const senderSig = sender.sign(ackCore(ack));
    const receipt: Receipt = {
      ack,
      sender: sender.id,
      receiver: receiver.id,
      senderSig,
      receiverSig,
    };
    receipts.push(receipt);

    if (walkaway) {
      // Exposure is exactly this one increment: the sender parted with the
      // bytes but holds no receiver acknowledgment for them. Record it as a
      // first-class reputation event, not an off-book loss.
      unsignedExposureBytes = incBytes;
      sender.ledger.append("reputation_event", ts, {
        kind: "unsigned_delivery",
        counterparty: receiver.id,
        cid,
        exposureBytes: incBytes,
        note: "receiver took increment without signing",
      });
      // The unsigned increment is NOT counted toward billable postage — you bill
      // what was agreed, and this was not agreed.
      moved -= incBytes;
    } else {
      const body = receiptBody(receipt);
      sender.ledger.append("receipt", ts, body);
      receiver.ledger.append("receipt", ts, body);
    }
  }

  return { receipts, total: moved, unsignedExposureBytes };
}

export function receiptBody(r: Receipt): Json {
  return {
    ack: ackCore(r.ack),
    sender: r.sender,
    receiver: r.receiver,
    senderSig: r.senderSig,
    receiverSig: r.receiverSig,
  };
}

/** Verify a receipt's two signatures against pinned peer keys held by `checker`. */
export function verifyReceipt(checker: Actor, r: Receipt): boolean {
  const core = ackCore(r.ack);
  const senderKey = checker.pinnedKeyFor(r.sender);
  const receiverKey = checker.pinnedKeyFor(r.receiver);
  if (!senderKey) return false;
  if (!verifyJson(senderKey, core, r.senderSig)) return false;
  if (r.receiverSig === null) return false; // unsigned increment does not verify
  if (!receiverKey) return false;
  return verifyJson(receiverKey, core, r.receiverSig);
}

export interface Reconciliation {
  ok: boolean;
  customerTotal: number;
  providerTotal: number;
  errors: string[];
}

/**
 * Reconcile the two ledgers' receipt rows. They must contain the same co-signed
 * acks, every embedded signature must verify, and the summed byte totals must be
 * identical. This is what exposes a party that altered a byte count in its own
 * copy: the counterparty signature no longer covers the altered ack.
 */
export function reconcile(customer: Actor, provider: Actor): Reconciliation {
  const errors: string[] = [];
  const cust = customer.ledger.entries().filter((e) => e.type === "receipt");
  const prov = provider.ledger.entries().filter((e) => e.type === "receipt");

  if (cust.length !== prov.length) {
    errors.push(`receipt count differs: customer ${cust.length} vs provider ${prov.length}`);
  }

  const sum = (rows: typeof cust): number =>
    rows.reduce((acc, e) => acc + Number((e.body as { ack: { incBytes: number } }).ack.incBytes), 0);

  // Every receipt on both sides must carry two valid signatures.
  const checkRows = (rows: typeof cust, checker: Actor, who: string): void => {
    for (const e of rows) {
      const b = e.body as unknown as Receipt;
      if (!verifyReceipt(checker, { ack: b.ack, sender: b.sender, receiver: b.receiver, senderSig: b.senderSig, receiverSig: b.receiverSig })) {
        errors.push(`${who} receipt seq ${e.seq} (${b.ack.cid} @${b.ack.rangeStart}) failed signature check`);
      }
    }
  };
  checkRows(cust, customer, "customer-side");
  checkRows(prov, provider, "provider-side");

  const customerTotal = sum(cust);
  const providerTotal = sum(prov);
  if (customerTotal !== providerTotal) {
    errors.push(`totals differ: customer ${customerTotal} vs provider ${providerTotal}`);
  }

  return { ok: errors.length === 0, customerTotal, providerTotal, errors };
}

File: alpha/experiments/item-store/src/receipts.rs
//! Transfer receipts: signed acknowledgments of bytes crossing the boundary
//! between two actors. Postage is charged by weight (bytes), not by trips —
//! "meter the boundary, not the machine".
//!
//! Two modes, selected per transfer by the social-trust layer:
//! - [`ReceiptMode::Bilateral`] — co-signed by both parties; self-contained and
//!   third-party-verifiable (the co-attested form the deferred capital layer
//!   E11–E14 will require).
//! - [`ReceiptMode::Unilateral`] — provider-signed only; an "our-side
//!   measurement", valid by the trust relationship but NOT co-attested, so its
//!   provenance is weaker.
//!
//! A receipt's signatures are taken over its own content (not a ledger
//! position), so the identical receipt can be embedded in both parties' ledgers
//! and cross-checked; altering a byte count in either copy breaks the embedded
//! signatures.
//!
//! Ports `item-storage-protocol-standalone/src/receipt.ts` and adds the
//! unilateral mode + the mode-selection seam.

use std::collections::BTreeMap;

use ed25519_dalek::VerifyingKey;
use serde::{Deserialize, Serialize};

use crate::canonical::to_canonical_bytes;
use crate::crypto::{sha256_hex, verify_message, Keypair};

/// Which way the bytes crossed the boundary.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Direction {
    /// Customer → provider.
    Upload,
    /// Provider → customer.
    Download,
}

/// How a receipt is attested.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum ReceiptMode {
    /// Provider-signed only — an our-side measurement, valid by trust.
    Unilateral,
    /// Co-signed by both parties — third-party-verifiable.
    Bilateral,
}

/// The signed content of a receipt (everything the signatures are taken over).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ReceiptCore {
    /// Transfer direction.
    pub direction: Direction,
    /// The content address transferred.
    pub cid: String,
    /// First byte of this increment (inclusive).
    pub byte_start: usize,
    /// Last byte of this increment (exclusive).
    pub byte_end: usize,
    /// Bytes in this increment (`byte_end - byte_start`).
    pub bytes: usize,
    /// Running total transferred so far in this exchange.
    pub running_total: usize,
    /// The day the transfer occurred (byte-day accounting uses this).
    pub day: u64,
    /// The party receiving the bytes.
    pub receiver_id: String,
    /// The party sending the bytes.
    pub sender_id: String,
}

impl ReceiptCore {
    /// Build a receipt core; `bytes` is derived from the byte range so it cannot
    /// disagree with it.
    #[must_use]
    pub fn new(
        direction: Direction,
        cid: &str,
        byte_range: (usize, usize),
        running_total: usize,
        day: u64,
        receiver_id: &str,
        sender_id: &str,
    ) -> Self {
        let (byte_start, byte_end) = byte_range;
        Self {
            direction,
            cid: cid.to_owned(),
            byte_start,
            byte_end,
            bytes: byte_end - byte_start,
            running_total,
            day,
            receiver_id: receiver_id.to_owned(),
            sender_id: sender_id.to_owned(),
        }
    }

    /// The content hash the receipt's signatures are taken over.
    #[must_use]
    pub fn content_hash(&self) -> String {
        sha256_hex(&to_canonical_bytes(self))
    }
}

/// A transfer receipt: a core, its content hash, its mode, and the signatures
/// gathered over the content hash (keyed by signer id).
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Receipt {
    core: ReceiptCore,
    content_hash: String,
    mode: ReceiptMode,
    sigs: BTreeMap<String, String>,
}

impl Receipt {
    /// Reconstruct a receipt from its parts — e.g. read back off the wire or out
    /// of a ledger body. (`verify_*` re-derive the content hash, so a
    /// reconstruction with an altered core will not pass verification.)
    #[must_use]
    pub fn from_parts(
        core: ReceiptCore,
        content_hash: String,
        mode: ReceiptMode,
        sigs: BTreeMap<String, String>,
    ) -> Self {
        Self {
            core,
            content_hash,
            mode,
            sigs,
        }
    }

    /// The signed core.
    #[must_use]
    pub fn core(&self) -> &ReceiptCore {
        &self.core
    }

    /// The stored content hash.
    #[must_use]
    pub fn content_hash(&self) -> &str {
        &self.content_hash
    }

    /// The attestation mode.
    #[must_use]
    pub fn mode(&self) -> ReceiptMode {
        self.mode
    }

    /// Bytes metered by this receipt.
    #[must_use]
    pub fn bytes(&self) -> usize {
        self.core.bytes
    }

    /// The signatures gathered, keyed by signer id.
    #[must_use]
    pub fn sigs(&self) -> &BTreeMap<String, String> {
        &self.sigs
    }

    /// Whether the stored content hash still matches the core (tamper check).
    fn core_matches(&self) -> bool {
        self.core.content_hash() == self.content_hash
    }

    /// Fully valid as a bilateral receipt: both parties signed the content and
    /// both signatures verify under their pinned keys.
    #[must_use]
    pub fn verify_bilateral(&self, keyring: &BTreeMap<String, VerifyingKey>) -> bool {
        if self.mode != ReceiptMode::Bilateral || !self.core_matches() {
            return false;
        }
        let (Some(receiver_sig), Some(sender_sig)) = (
            self.sigs.get(&self.core.receiver_id),
            self.sigs.get(&self.core.sender_id),
        ) else {
            return false;
        };
        let (Some(receiver_key), Some(sender_key)) = (
            keyring.get(&self.core.receiver_id),
            keyring.get(&self.core.sender_id),
        ) else {
            return false;
        };
        verify_message(receiver_key, &self.content_hash, receiver_sig)
            && verify_message(sender_key, &self.content_hash, sender_sig)
    }

    /// Valid as a unilateral (provider-signed) measurement: every signature
    /// present verifies under a pinned key. Weaker than bilateral — single
    /// party, not co-attested.
    #[must_use]
    pub fn verify_unilateral(&self, keyring: &BTreeMap<String, VerifyingKey>) -> bool {
        if self.mode != ReceiptMode::Unilateral || !self.core_matches() || self.sigs.is_empty() {
            return false;
        }
        self.sigs.iter().all(|(id, sig)| {
            keyring
                .get(id)
                .is_some_and(|key| verify_message(key, &self.content_hash, sig))
        })
    }

    /// Whether both parties acknowledged (distinguishes a walkaway, which
    /// carries no signatures).
    #[must_use]
    pub fn is_acknowledged(&self) -> bool {
        self.sigs.contains_key(&self.core.receiver_id)
            && self.sigs.contains_key(&self.core.sender_id)
    }

    /// Whether this receipt is co-attested (bilateral, both parties present).
    /// A unilateral receipt is never co-attested.
    #[must_use]
    pub fn is_co_attested(&self) -> bool {
        self.mode == ReceiptMode::Bilateral
            && self.sigs.contains_key(&self.core.receiver_id)
            && self.sigs.contains_key(&self.core.sender_id)
    }
}

/// Inputs the social-trust policy weighs when choosing a receipt mode.
pub struct TransferContext {
    /// Bytes about to cross the boundary.
    pub bytes: usize,
    /// Trust distance to the counterparty, if the trust layer supplied one.
    pub trust_distance: Option<u32>,
}

/// Select the receipt mode for a transfer.
///
/// `SEAM:` the real policy is the social-trust layer deciding Unilateral vs
/// Bilateral per transfer (size / sensitivity / trust-distance), reusing the
/// same trust-distance primitive as the forum's subjective consensus. v0
/// returns the configured `default`; the policy hook lands in a later spike.
#[must_use]
pub fn select_mode(_ctx: &TransferContext, default: ReceiptMode) -> ReceiptMode {
    // SEAM: trust-distance-driven selection goes here.
    default
}

/// Build a bilateral receipt (free function matching the oracle's `makeReceipt`).
/// The receiver signs first (acknowledging receipt); the sender only
/// countersigns an acknowledged transfer. A walkaway is modelled by passing
/// `None` for the receiver — the result carries no signatures.
#[must_use]
pub fn make_bilateral_receipt(
    core: ReceiptCore,
    receiver_key: Option<&Keypair>,
    sender_key: &Keypair,
) -> Receipt {
    let content_hash = core.content_hash();
    let mut sigs = BTreeMap::new();
    if let Some(receiver_key) = receiver_key {
        sigs.insert(
            core.receiver_id.clone(),
            receiver_key.sign_message(&content_hash),
        );
        sigs.insert(
            core.sender_id.clone(),
            sender_key.sign_message(&content_hash),
        );
    }
    Receipt::from_parts(core, content_hash, ReceiptMode::Bilateral, sigs)
}

/// Build a unilateral receipt: the provider signs its own-side measurement.
#[must_use]
pub fn make_unilateral_receipt(
    core: ReceiptCore,
    provider_id: &str,
    provider_key: &Keypair,
) -> Receipt {
    let content_hash = core.content_hash();
    let mut sigs = BTreeMap::new();
    sigs.insert(
        provider_id.to_owned(),
        provider_key.sign_message(&content_hash),
    );
    Receipt::from_parts(core, content_hash, ReceiptMode::Unilateral, sigs)
}

#[cfg(test)]
mod tests {
    use super::{
        make_bilateral_receipt, make_unilateral_receipt, Direction, Receipt, ReceiptCore,
        ReceiptMode,
    };
    use crate::crypto::derive_keypair;
    use crate::identity::derive_id;
    use ed25519_dalek::VerifyingKey;
    use std::collections::BTreeMap;

    fn parties() -> (
        String,
        crate::crypto::Keypair,
        String,
        crate::crypto::Keypair,
    ) {
        let receiver = derive_keypair("m", "receiver");
        let sender = derive_keypair("m", "sender");
        let rid = derive_id(&receiver.verifying_key());
        let sid = derive_id(&sender.verifying_key());
        (rid, receiver, sid, sender)
    }

    fn ring(entries: &[(&str, &crate::crypto::Keypair)]) -> BTreeMap<String, VerifyingKey> {
        entries
            .iter()
            .map(|(id, kp)| ((*id).to_owned(), kp.verifying_key()))
            .collect()
    }

    #[test]
    fn bilateral_receipt_verifies_and_detects_tampering() {
        let (rid, receiver, sid, sender) = parties();
        let ring = ring(&[(&rid, &receiver), (&sid, &sender)]);
        let core = ReceiptCore::new(Direction::Upload, "cid", (0, 100), 100, 1, &rid, &sid);
        let receipt = make_bilateral_receipt(core, Some(&receiver), &sender);
        assert!(receipt.verify_bilateral(&ring));
        assert!(receipt.is_acknowledged());

        let mut tampered_core = receipt.core().clone();
        tampered_core.bytes += 1;
        let tampered = Receipt::from_parts(
            tampered_core,
            receipt.content_hash().to_owned(),
            receipt.mode(),
            receipt.sigs().clone(),
        );
        assert!(
            !tampered.verify_bilateral(&ring),
            "tamper breaks the content hash"
        );
    }

    #[test]
    fn walkaway_receipt_carries_no_signatures() {
        let (rid, receiver, sid, sender) = parties();
        let core = ReceiptCore::new(Direction::Download, "cid", (0, 100), 100, 1, &rid, &sid);
        let _ = &receiver;
        let receipt = make_bilateral_receipt(core, None, &sender);
        assert!(!receipt.is_acknowledged());
        assert!(receipt.sigs().is_empty());
    }

    // The following are mutation-resistance tests (E86): they pin the boolean
    // structure of the verify predicates that a happy-path suite leaves free.

    #[test]
    fn bilateral_with_one_bad_signature_fails() {
        // Exactly one of the two signatures is invalid — the AND in
        // verify_bilateral must reject (a lone valid sig is not enough).
        let (rid, receiver, sid, sender) = parties();
        let ring = ring(&[(&rid, &receiver), (&sid, &sender)]);
        let core = ReceiptCore::new(Direction::Upload, "cid", (0, 100), 100, 1, &rid, &sid);
        let good = make_bilateral_receipt(core, Some(&receiver), &sender);

        let mut sigs = good.sigs().clone();
        // Overwrite the sender's signature with one over a different message.
        sigs.insert(sid.clone(), sender.sign_message("some other message"));
        let one_bad = Receipt::from_parts(
            good.core().clone(),
            good.content_hash().to_owned(),
            ReceiptMode::Bilateral,
            sigs,
        );
        assert!(
            !one_bad.verify_bilateral(&ring),
            "one valid + one invalid signature must not verify bilaterally",
        );
    }

    #[test]
    fn bilateral_with_only_one_signature_is_not_acknowledged_or_co_attested() {
        let (rid, receiver, sid, sender) = parties();
        let core = ReceiptCore::new(Direction::Upload, "cid", (0, 100), 100, 1, &rid, &sid);
        let good = make_bilateral_receipt(core, Some(&receiver), &sender);

        // Drop the sender's signature — only the receiver's remains.
        let mut sigs = good.sigs().clone();
        sigs.remove(&sid);
        let one_sig = Receipt::from_parts(
            good.core().clone(),
            good.content_hash().to_owned(),
            ReceiptMode::Bilateral,
            sigs,
        );
        assert!(
            !one_sig.is_acknowledged(),
            "one signature is not acknowledged"
        );
        assert!(
            !one_sig.is_co_attested(),
            "one signature is not co-attested"
        );
    }

    #[test]
    fn acknowledged_bilateral_receipt_is_co_attested() {
        let (rid, receiver, sid, sender) = parties();
        let core = ReceiptCore::new(Direction::Upload, "cid", (0, 100), 100, 1, &rid, &sid);
        let receipt = make_bilateral_receipt(core, Some(&receiver), &sender);
        assert!(
            receipt.is_co_attested(),
            "a fully-signed bilateral receipt is co-attested",
        );
    }

    #[test]
    fn unilateral_rejection_paths() {
        let (rid, receiver, sid, sender) = parties();
        let ring = ring(&[(&rid, &receiver), (&sid, &sender)]);
        let core = ReceiptCore::new(Direction::Upload, "cid", (0, 100), 100, 1, &rid, &sid);

        // A valid unilateral (provider = sender here) measurement verifies.
        let good = make_unilateral_receipt(core.clone(), &sid, &sender);
        assert!(good.verify_unilateral(&ring));

        // (a) tampered core: the content hash no longer matches -> reject.
        let mut tampered_core = good.core().clone();
        tampered_core.bytes += 1;
        let tampered = Receipt::from_parts(
            tampered_core,
            good.content_hash().to_owned(),
            ReceiptMode::Unilateral,
            good.sigs().clone(),
        );
        assert!(!tampered.verify_unilateral(&ring), "tampered core rejected");

        // (b) no signatures: an unsigned measurement is not valid.
        let empty = Receipt::from_parts(
            good.core().clone(),
            good.content_hash().to_owned(),
            ReceiptMode::Unilateral,
            std::collections::BTreeMap::new(),
        );
        assert!(
            !empty.verify_unilateral(&ring),
            "empty-sig unilateral rejected"
        );

        // (c) wrong mode: a Bilateral-mode receipt is not unilaterally valid.
        let bilateral = make_bilateral_receipt(core, Some(&receiver), &sender);
        assert!(
            !bilateral.verify_unilateral(&ring),
            "bilateral is not unilaterally valid",
        );
    }
}
